### PR TITLE
Fix Meridian RGB indicators for varying hardware across PCB batches. 

### PR DESCRIPTION
**Pull request by holtenc**
Two different LED parts were used by the manufacturer creating a firmware issue. This update helps to correct this by making two different firmware versions. A small change was made to drivers/chibios/ws2812.c to make this possible. 

WS2812 driver was also changed from SPI to bitbang due to an apparent issue with SPI introduced since the Meridian firmware was initially merged. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

https://github.com/qmk/qmk_firmware/issues/11755

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/primekb/meridian/meridian.c [hunk 1]
```diff
@@ -35,19 +35,19 @@ void keyboard_post_init_user(void) {
 bool led_update_kb(led_t led_state) {
```
Comment (newtonapple) [inline keyboards/primekb/meridian/meridian.c:35]:
```suggestion
__attribute__((weak))
bool led_update_kb(led_t led_state) {
```

File: drivers/chibios/ws2812.c [hunk 2]
```diff
@@ -43,11 +43,21 @@
 // These are the timing constraints taken mostly from the WS2812 datasheets
 // These are chosen to be conservative and avoid problems rather than for maximum throughput
 
+#ifndef T1H
```
Comment (drashna) [inline drivers/chibios/ws2812.c:46]: This should be done as a separate PR.  

Please pull this change out, and submit it as such.

File: keyboards/primekb/meridian/v1.1/readme.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+![Meridian](https://cdn.shopify.com/s/files/1/2016/1211/files/IMG_0934.JPG?v=1612649071)
+
+* Keyboard Maintainer: [Holten Campbell](https://github.com/holtenc)
+* Hardware Supported: STM32F072CBT6 with KTR-1010RGBC-1812-4 LEDs
+
+V1.1 PCBs are generally identified by:
+* Having a holographic sticker on the front reading "QC3" or "QC5".
+* By visually identifying which LEDs you have in reference to the abvove image. 
+
+Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/primekb/meridian/v1.1/readme.md:12]: Please add instructions on how to enter the bootloader mode.

File: keyboards/primekb/meridian/v1/readme.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+![Meridian](https://cdn.shopify.com/s/files/1/2016/1211/files/IMG_0934.JPG?v=1612649071)
+
+* Keyboard Maintainer: [Holten Campbell](https://github.com/holtenc)
+* Hardware Supported: STM32F072CBT6 with WS2812C LEDs
+
+V1 PCBs are generally identified by: 
+* Having a holographic sticker on the front reading "QC1"
+* By visually identifying which LEDs you have in reference to the abvove image. 
+
+Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/primekb/meridian/v1/readme.md:12]: Please add instructions on how to enter the bootloader mode.

File: keyboards/primekb/meridian/meridian.c (hunk 1 above)
Comment (drashna) [reply keyboards/primekb/meridian/meridian.c:35]: @newtonapple This change isn't actually needed.  First, it's the keyboard level function, so ... it shouldn't be replaced.

Second, the `res` bool allows users to replace/disable this functionality, by returning false from `led_update_user`

File: keyboards/primekb/meridian/meridian.c [hunk 5]
```diff
@@ -35,19 +35,19 @@ void keyboard_post_init_user(void) {
...
     bool res = led_update_user(led_state);
     if (res) {
- //       writePin(B12, !led_state.caps_lock);  //Un-comment this line to enable in-switch capslock indicator
+        writePin(B12, !led_state.caps_lock);
     if (led_state.caps_lock) {
-        rgblight_setrgb_at(0, 255, 0, 0); //green
+        rgblight_setrgb_at(255, 255, 255, 0);
```
Comment (drashna) [inline keyboards/primekb/meridian/meridian.c:40]: I think that the rgblight layers feature will handle this better, as it runs the code from `rgblight_set` itself, so it is called after animations are written to the array, but right before they're sent to the LEDs.

**Review by drashna — CHANGES_REQUESTED** [review]

File: drivers/chibios/ws2812.c (hunk 2 above)
Comment (holtenc) [reply drivers/chibios/ws2812.c]: Done. I'll submit another PR

File: keyboards/primekb/meridian/v1.1/config.h [hunk 6]
```diff
@@ -0,0 +1,6 @@
+//Data Transfer Time for KTR-1010RGBC-1812-4 LEDs as per datasheet spec.
```
Comment (zvecr) [inline keyboards/primekb/meridian/v1.1/config.h:1]:
```suggestion
/*
Copyright 2020 Holten Campbell
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#pragma once
//Data Transfer Time for KTR-1010RGBC-1812-4 LEDs as per datasheet spec. 
```

File: keyboards/primekb/meridian/v1.1/rules.mk [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
+
+#DEFAULT_FOLDER = primekb/meridian/v1.1
```
Comment (zvecr) [inline keyboards/primekb/meridian/v1.1/rules.mk:24]:
```suggestion
```

File: keyboards/primekb/meridian/v1/config.h [hunk 8]
```diff
@@ -0,0 +1,6 @@
+//Data Transfer Time for WS2812C LEDs as per datasheet spec.
```
Comment (zvecr) [inline keyboards/primekb/meridian/v1/config.h:1]:
```suggestion
/*
Copyright 2020 Holten Campbell
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#pragma once
//Data Transfer Time for WS2812C LEDs as per datasheet spec.
```

File: keyboards/primekb/meridian/v1/rules.mk [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
+
+#DEFAULT_FOLDER = primekb/meridian/v1.1
```
Comment (zvecr) [inline keyboards/primekb/meridian/v1/rules.mk:24]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Also `v1.1` is currently an invalid keyboard name so would have to be something like `v1_1`

**Comment by holtenc** [comment]
build errors in continuous-integration/travis-ci/pr are due to my changes in drivers/chibios/ws2812.c not being present yet.

File: keyboards/primekb/meridian/meridian.c [hunk 10]
```diff
@@ -23,46 +23,42 @@ void keyboard_pre_init_kb(void){
...
-   if (led_state.scroll_lock) {          
-        rgblight_setrgb_at(255, 0, 0, 2); //red
-    } else {
-        rgblight_setrgb_at(0, 0, 0, 2);
-    }
-}
-    return res;
+    rgblight_set_layer_state(0, led_state.caps_lock);
+    rgblight_set_layer_state(1, led_state.num_lock);
+    rgblight_set_layer_state(2, led_state.scroll_lock);
```
Comment (newtonapple) [inline keyboards/primekb/meridian/meridian.c:62]: Should these be inside of the `if(res) {}` block instead?

File: keyboards/primekb/meridian/rules.mk [hunk 11]
```diff
@@ -17,7 +17,8 @@ BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
 RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
 BLUETOOTH_ENABLE = no       # Enable Bluetooth
 AUDIO_ENABLE = no           # Audio output
-WS2812_DRIVER = spi
 
 # Enter lower-power sleep mode when on the ChibiOS idle thread
 OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
+
+DEFAULT_FOLDER = primekb/meridian/v1.1
```
Comment (zvecr) [inline keyboards/primekb/meridian/rules.mk:24]: Should be the old folder to maintain backwards compatibility
```suggestion
DEFAULT_FOLDER = primekb/meridian/v1
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Also as a passing comment, the rev1 now using the bitbang driver seems like a step backwards. And use of rgblight layers at the keyboard level seems a tad meh.

File: keyboards/primekb/meridian/rules.mk (hunk 11 above)
Comment (holtenc) [reply keyboards/primekb/meridian/rules.mk:24]: I chose 1.1 as the default since there about 5x as many 1.1 PCBs available than 1.0

Switch to bitbang for 2 reasons: 1) there seemed to be a bug introduced in the SPI driver after the initial Meridian commit, as reported here: https://github.com/qmk/qmk_firmware/issues/11755. And 2)  So far as I can tell I can't define data transfer times through the SPI driver (ie T0H, T1H, T0L, T1L), which is a must. 

@drashna said to use rgblight layers, so I changed it. Guys I don't care how it's implemented so long as it works and it gets approved/merged. Lets all get on the same page, tell me how you want this part done and I'll do it. 

Thanks guys

**Comment by holtenc** [comment]
@drashna @zvecr 
Do you guys have any further revisions that need to be made? I can get that PR for /drivers/chibios/ws2812.c submitted as soon as this side is taken care of.

File: keyboards/primekb/meridian/meridian.c [hunk 12]
```diff
@@ -23,46 +23,42 @@ void keyboard_pre_init_kb(void){
     keyboard_pre_init_user();
 }
 
-//Initialize all RGB indicators to 'off'
+//Refernce the following QMK documentation for implementing Lighting Layers and layer change indication:
+// https://beta.docs.qmk.fm/using-qmk/hardware-features/lighting/feature_rgblight#lighting-layers
```
Comment (drashna) [inline keyboards/primekb/meridian/meridian.c:27]:
```suggestion
// https://docs.qmk.fm/#/feature_rgblight?id=lighting-layers
```

File: keyboards/primekb/meridian/meridian.c (hunk 10 above)
Comment (drashna) [reply keyboards/primekb/meridian/meridian.c:62]: They should be, IMO, yeah. Otherwise, users can't over ride the feature.

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Comment by holtenc** [comment]
So would it be best to just close this PR and re-submit? @zvecr @drashna 

from what I can tell the structure of halconf.h and chconf.h etc have had significant changes, which I attempted to update to but that just seems to have messed this PR up. 

please advise.

**Comment by drashna** [comment]
Well, updating them should be just running qmk chibosconf, I think it was. 

But you roll back commits and force push, or revert them, to fix issues.
